SingularityS3Uploader/src/main/java/com/hubspot/singularity/s3uploader/SingularityS3UploaderDriver.java

```diff
@@ -256,31 +268,35 @@ private int checkUploads() {
       }
     }
 
-    for (SingularityUploader uploader : toRetry) {
-      LOG.debug("Retrying immediate uploader {}", uploader);
-      performImmediateUpload(uploader);
+    for (S3UploadMetadata uploaderMetadata : toRetry) {
+      SingularityUploader uploader = metadataToimmediateUploader.remove(uploaderMetadata);
+      if (uploader != null) {
+        LOG.debug("Retrying immediate uploader {}", uploaderMetadata);
+        performImmediateUpload(uploader);
```

Comment:
For this retry loop, the uploader is removed from `metadataToImmeditateUploader` and when it's called to upload immediately, it'll be added to the `immediteUploadersFutures` map. During the next run of `checkUploads`, won't the uploader be missing from `metadataToImmediateUploader` and cause the check to skip adding the uploadedFiles to the total count? Instead it'll be moved to the `toRemove` list and hit the 30 second check

Reply:
good catch. It ends up working out that it still retries, but that should be a .get so we can do the full retry -> toRemove loop . Fixed